Implement a ping pong game according to the template

loganstillings opened the pull request:
### What
A basic ping pong game written with HTML and JavaScript.

Features:
-   The game covers the whole browser and is resizable.
-   The keyboard controls both players. W and S control the red player, Up and Down control the blue player.
-   Pressing `SPACE` will start and pause the game.
-   The ball bounces off walls and paddles. Vertical ball speed is calculated based on where the ball hits the paddle in relation to the paddle's center.
-   Scoring is logged to the console as well as displayed on the game board.
-   Ball is served in a random direction.
-   State functions allow the user to reset, start, and pause the game from the console.

### Why
To create an enjoyable front-end game of ping pong that also exemplifies modular, clean code. This is also meant to showcase my ability to follow technical instructions and implement new skills quickly.

### How
The game was written entirely with HTML and Javascript using the provided template. Minimal changes were added to the app.js file (though my IDE reformatted the code, thus making the PR diff seem large). The app.js changes were to make the ball round and to add the scoring overlay via fillText in the app.update function. Keybindings and Node creation of the Ball and Players is in the index.html app.OnInit method. Separate state functions for resetting, starting, and pausing are also defined there. The entities.js file contains logic for Node, Player, and Ball classes including how to update their location per frame and collision handling. 

### Demo
[Video](https://recordit.co/4F6CDQnALO) (please note the video quality makes the game appear choppier than it really is). The Video can also be downloaded from the repository (demo.mp4)

loganstillings committed:
Create paddle nodes and move paddles on player input

loganstillings committed:
Add ball and basic collision/motion logic

loganstillings committed:
Refactor to handle update logic per entity

loganstillings committed:
Add restart logic after point is scored

loganstillings committed:
Implement resizing

loganstillings committed:
Add state functions for pause, reset, and start

loganstillings committed:
Make entities.js file agnostic of global constants

loganstillings committed:
Invert game to be vertically aligned

loganstillings committed:
Randomize "serve" direction

loganstillings committed:
Display Score with fillText

loganstillings committed:
Make ball round

loganstillings committed:
Update score font and opacity

loganstillings committed:
Fix bug where ball would get caught within paddle

loganstillings committed:
Implement vertical ball speed on return
calculated based on where on the paddle the ball is hitting

loganstillings committed:
Update README and add demo Video

demiculus commented:
@loganstillings well done can you do a couple edits on top of this?

1. Use early breaks: 
    - https://jamesmonger.com/2019/08/06/return-early-return-often.html
    - https://raquo.net/2016/01/guard-statement-swift/
2. Reduce the length of the functions and reduce the amount of indents each function has. You can create more function on wherever you see necessary.
3. Add a fun mechanic. Anything you can think of that would make the game more fun.

Esqarrouth commented:
Umm, have you done changes since last comment? I don't see it?

loganstillings commented:
> Umm, have you done changes since last comment? I don't see it?

I'm still working on the changes mentioned by demiculus, I should have those changes ready soon. I just requested reviews as per the instructions. Sorry for the confusion.

loganstillings committed:
Refactor to use shorter functions

loganstillings committed:
Add Powerups

loganstillings committed:
Refactor with shorter methods
Created some additional functions to declutter

loganstillings committed:
Update README and demo with powerup functionality

loganstillings commented:
@demiculus @goktugyil Hey guys, I've made some changes based on the feedback from Demi. I refactored the code to use more early breaks and shorter methods (by adding more methods overall). To make the game more fun, I added power-ups to the game (paddle growth, paddle speed boost, and freezing the opposing player). Please take a look and let me know if you have more feedback. Thanks!
